src/main/java/bio/terra/common/kubernetes/KubeService.java

```diff
@@ -0,0 +1,190 @@
+ * that fail when a pod goes away. The current convention is to use the pod name as the stairway
+ * name. Then the service can compare the list of stairways it thinks should be running, compare
+ * with list of active pods, and recover anything that is no longer active.
+ */
+@SuppressFBWarnings(
+    value = "DMI_HARDCODED_ABSOLUTE_FILENAME",
+    justification = "The K8s namespace file is a valid absolute filename")
+public class KubeService {
+  private static final Logger logger = LoggerFactory.getLogger(KubeService.class);
+
+  // Location in the container where Kubernetes stores service account and
+  // namespace information. Kubernetes mints a service account for the container (pod?)
+  // that can be used to make requests of Kubernetes.
+  static final String KUBE_DIR = "/var/run/secrets/kubernetes.io/serviceaccount";
+  static final String KUBE_NAMESPACE_FILE = KUBE_DIR + "/namespace";
+
+  private final String podName;
+  private final boolean inKubernetes;
+  private final String podNameFilter;
+  private final String namespace;
+  private final AtomicBoolean isShutdown;
+
+  private KubePodListener podListener;
+  private Thread podListenerThread;
+
+  /**
+   * @param podName Name of the Kubernetes pod we are running in. If we are not in a pod, this
+   *     defaults to a constant string in application properties. We cannot easily find our own pod
+   *     name, so components must plumb this in using settings in the helm charts.
+   * @param inKubernetes Used to denote that we are running in the Kubernetes environment. The
+   *     service can be called and provides reasonable answers when inKubernetes is false.
+   * @param podNameFilter Filter to apply to pod names returned by Kubernetes API to know which pods
+   *     we care about.
+   */
+  public KubeService(String podName, boolean inKubernetes, String podNameFilter) {
+    this.podName = podName;
+    this.inKubernetes = inKubernetes;
+    this.podNameFilter = podNameFilter;
+    if (inKubernetes) {
+      this.namespace = readFileIntoString(KUBE_NAMESPACE_FILE);
+    } else {
+      this.namespace = "nonamespace";
+    }
+    this.isShutdown = new AtomicBoolean(false);
+
+    logger.info(
+        "Kubernetes configuration: inKube: "
+            + inKubernetes
+            + "; namespace: "
+            + namespace
+            + "; podName: "
+            + podName);
+  }
+
+  /**
+   * Get a list of the API pods from Kubernetes. It is returned as a set to make probing for
+   * specific names easy.
+   *
+   * @return set of pod names containing the podNameFilter string; null if not in kubernetes
+   */
+  public Set<String> getPodList() {
+    Set<String> pods = new HashSet<>();
+    if (!inKubernetes) {
+      return pods;
+    }
+
+    try {
+      CoreV1Api api = makeCoreApi();
+      V1PodList list =
+          api.listNamespacedPod(namespace, null, null, null, null, null, null, null, null, null);
```

Comment:
Do we control this api? Can we overload `listNamespacePod`?

Reply:
No, we do not control the api.